// browser-use-convex-component/dist/client/index.js
export class BrowserUse {
    component;
    _apiKey;
    constructor(component, options) {
        this.component = component;
        this._apiKey = options?.BROWSER_USE_API_KEY;
    }
    get apiKey() {
        const key = this._apiKey ?? process.env.BROWSER_USE_API_KEY;
        if (!key) {
            throw new Error("Browser Use API key is required. Pass it as BROWSER_USE_API_KEY option or set the BROWSER_USE_API_KEY environment variable.");
        }
        return key;
    }
    // ------- Tasks -------
    async createTask(ctx, args) {
        return await ctx.runAction(this.component.tasks.create, {
            apiKey: this.apiKey,
            ...args,
        });
    }
    async getTask(ctx, args) {
        return await ctx.runQuery(this.component.tasks.get, {
            taskId: args.taskId,
        });
    }
    async listTasks(ctx, args) {
        return await ctx.runQuery(this.component.tasks.list, args ?? {});
    }
    async getTaskSteps(ctx, args) {
        return await ctx.runQuery(this.component.tasks.getSteps, {
            taskId: args.taskId,
        });
    }
    async fetchTaskStatus(ctx, args) {
        return await ctx.runAction(this.component.tasks.fetchStatus, {
            apiKey: this.apiKey,
            externalId: args.externalId,
        });
    }
    async fetchTaskDetail(ctx, args) {
        return await ctx.runAction(this.component.tasks.fetchDetail, {
            apiKey: this.apiKey,
            externalId: args.externalId,
        });
    }
    async stopTask(ctx, args) {
        return await ctx.runAction(this.component.tasks.stop, {
            apiKey: this.apiKey,
            externalId: args.externalId,
        });
    }
    // ------- Sessions -------
    async createSession(ctx, args) {
        return await ctx.runAction(this.component.sessions.create, {
            apiKey: this.apiKey,
            ...(args ?? {}),
        });
    }
    async getSession(ctx, args) {
        return await ctx.runQuery(this.component.sessions.get, {
            sessionId: args.sessionId,
        });
    }
    async listSessions(ctx, args) {
        return await ctx.runQuery(this.component.sessions.list, args ?? {});
    }
    async fetchSessionDetail(ctx, args) {
        return await ctx.runAction(this.component.sessions.fetchDetail, {
            apiKey: this.apiKey,
            externalId: args.externalId,
        });
    }
    async stopSession(ctx, args) {
        return await ctx.runAction(this.component.sessions.stop, {
            apiKey: this.apiKey,
            externalId: args.externalId,
        });
    }
    // ------- Profiles -------
    async createProfile(ctx, args) {
        return await ctx.runAction(this.component.profiles.create, {
            apiKey: this.apiKey,
            ...(args ?? {}),
        });
    }
    async listProfiles(ctx) {
        return await ctx.runAction(this.component.profiles.list, {
            apiKey: this.apiKey,
        });
    }
    async getProfile(ctx, args) {
        return await ctx.runAction(this.component.profiles.getProfile, {
            apiKey: this.apiKey,
            profileId: args.profileId,
        });
    }
    async updateProfile(ctx, args) {
        return await ctx.runAction(this.component.profiles.update, {
            apiKey: this.apiKey,
            ...args,
        });
    }
    async deleteProfile(ctx, args) {
        return await ctx.runAction(this.component.profiles.deleteProfile, {
            apiKey: this.apiKey,
            profileId: args.profileId,
        });
    }
}
