// browser-use-convex-component/src/component/_generated/component.ts
/* eslint-disable */
/**
 * Generated `ComponentApi` utility.
 *
 * THIS CODE IS AUTOMATICALLY GENERATED.
 *
 * To regenerate, run `npx convex dev`.
 * @module
 */

import type { FunctionReference } from "convex/server";

/**
 * A utility for referencing a Convex component's exposed API.
 *
 * Useful when expecting a parameter like `components.myComponent`.
 * Usage:
 * ```ts
 * async function myFunction(ctx: QueryCtx, component: ComponentApi) {
 *   return ctx.runQuery(component.someFile.someQuery, { ...args });
 * }
 * ```
 */
export type ComponentApi<Name extends string | undefined = string | undefined> =
  {
    profiles: {
      create: FunctionReference<
        "action",
        "internal",
        { apiKey: string; name?: string },
        { createdAt: string; id: string; name?: string },
        Name
      >;
      deleteProfile: FunctionReference<
        "action",
        "internal",
        { apiKey: string; profileId: string },
        null,
        Name
      >;
      getProfile: FunctionReference<
        "action",
        "internal",
        { apiKey: string; profileId: string },
        {
          cookieDomains: Array<string>;
          createdAt: string;
          id: string;
          lastUsedAt?: string;
          name?: string;
        },
        Name
      >;
      list: FunctionReference<
        "action",
        "internal",
        { apiKey: string },
        Array<{
          cookieDomains: Array<string>;
          createdAt: string;
          id: string;
          lastUsedAt?: string;
          name?: string;
        }>,
        Name
      >;
      update: FunctionReference<
        "action",
        "internal",
        { apiKey: string; name: string; profileId: string },
        { id: string; name?: string },
        Name
      >;
    };
    sessions: {
      create: FunctionReference<
        "action",
        "internal",
        {
          apiKey: string;
          browserScreenHeight?: number;
          browserScreenWidth?: number;
          profileId?: string;
          proxyCountryCode?: string;
          startUrl?: string;
        },
        { externalId: string; liveUrl?: string; sessionId: string },
        Name
      >;
      fetchDetail: FunctionReference<
        "action",
        "internal",
        { apiKey: string; externalId: string },
        any,
        Name
      >;
      get: FunctionReference<
        "query",
        "internal",
        { sessionId: string },
        any,
        Name
      >;
      list: FunctionReference<
        "query",
        "internal",
        { limit?: number; status?: "active" | "stopped" },
        any,
        Name
      >;
      stop: FunctionReference<
        "action",
        "internal",
        { apiKey: string; externalId: string },
        null,
        Name
      >;
    };
    tasks: {
      create: FunctionReference<
        "action",
        "internal",
        {
          allowedDomains?: Array<string>;
          apiKey: string;
          flashMode?: boolean;
          highlightElements?: boolean;
          judge?: boolean;
          judgeGroundTruth?: string;
          judgeLlm?: string;
          llm?: string;
          maxSteps?: number;
          metadata?: any;
          secrets?: any;
          sessionId?: string;
          skillIds?: Array<string>;
          startUrl?: string;
          structuredOutput?: any;
          systemPromptExtension?: string;
          task: string;
          thinking?: boolean;
          vision?: boolean | "auto";
        },
        { externalId: string; sessionId: string; taskId: string },
        Name
      >;
      fetchDetail: FunctionReference<
        "action",
        "internal",
        { apiKey: string; externalId: string },
        any,
        Name
      >;
      fetchStatus: FunctionReference<
        "action",
        "internal",
        { apiKey: string; externalId: string },
        { isSuccess?: boolean; output?: string; status: string },
        Name
      >;
      get: FunctionReference<
        "query",
        "internal",
        { taskId: string },
        any,
        Name
      >;
      getSteps: FunctionReference<
        "query",
        "internal",
        { taskId: string },
        any,
        Name
      >;
      list: FunctionReference<
        "query",
        "internal",
        {
          limit?: number;
          sessionId?: string;
          status?:
            | "created"
            | "started"
            | "paused"
            | "finished"
            | "stopped"
            | "failed";
        },
        any,
        Name
      >;
      stop: FunctionReference<
        "action",
        "internal",
        { apiKey: string; externalId: string },
        null,
        Name
      >;
    };
  };
